confd/warp/warp.go

```diff
@@ -0,0 +1,228 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read key %s from etcd", key)
+	}
+	external := RawData{}
+	err = json.Unmarshal([]byte(resp.Node.Value), &external)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshall json from etcd")
+	}
+
+	return external, nil
+}
+
+func createHref(dogu RawData) string {
+	// remove namespace
+	parts := strings.Split(dogu["Name"].(string), "/")
+	return "/" + parts[len(parts)-1]
+}
+
+func createDoguEntry(element RawData) Entry {
+	return Entry{
+		DisplayName: element["DisplayName"].(string),
+		Href:        createHref(element),
+		Title:       element["Description"].(string),
+		Target:      "self",
+	}
+}
+
+func createExternalEntry(element RawData) Entry {
+	return Entry{
+		DisplayName: element["DisplayName"].(string),
+		Href:        element["URL"].(string),
+		Title:       element["Description"].(string),
+		Target:      "external",
+	}
+}
+
+func filterByTag(dogus []RawData, tag string) []RawData {
+	filtered := []RawData{}
+	for _, raw := range dogus {
+		if raw["Tags"] != nil {
+			tags := raw["Tags"].([]interface{})
+			if tags != nil && Contains(tags, tag) {
+				filtered = append(filtered, raw)
+			}
+		}
+	}
+	return filtered
+}
+
+func (destination *Categories) insertCategories(newCategories Categories) {
+	for _, newCategory := range newCategories {
+		destination.insertCategory(newCategory)
+	}
+}
+
+func (destination *Categories) insertCategory(newCategory *Category) {
+	for _, category := range *destination {
+		if category.Title == newCategory.Title {
+			category.Entries = append(category.Entries, newCategory.Entries...)
+			return
+		}
+	}
+	*destination = append(*destination, newCategory)
+}
+
+// JSONWriter converts the data to a json
+func jsonWriter(target string, data interface{}) error {
+	bytes, err := json.Marshal(data)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal data to json")
+	}
+
+	return ioutil.WriteFile(target, bytes, 0755)
+}
+
+func execute(configuration Configuration, kapi client.KeysAPI) {
+	reader := ConfigReader{
+		kapi:          kapi,
+		configuration: configuration,
+	}
+	categories, err := reader.readFromConfig(configuration, kapi)
+	if err != nil {
+		log.Println("Error durring read", err)
+	}
+	log.Printf("all found categories: %i", categories)
+	jsonWriter(configuration.Target, categories)
+}
+
+func watch(source Source, kapi client.KeysAPI, execChannel chan Source) {
+	watcherOpts := client.WatcherOptions{AfterIndex: 0, Recursive: true}
+	watcher := kapi.Watcher(source.Path, &watcherOpts)
+	for {
+		resp, err := watcher.Next(context.Background())
+		if err != nil {
+			// TODO: execute before watch start again? wait to reduce load, in case of unrecoverable error?
+			watch(source, kapi, execChannel)
+		} else {
+			action := resp.Action
+			log.Printf("%s changed, action=%s", resp.Node.Key, action)
+			execChannel <- source
+		}
+	}
+}
+
+func Run(configuration Configuration, kapi client.KeysAPI) {
```

Comment:
**[checkstyle]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function Run should have comment or be unexported